packages/relay/src/routes/vote.ts

```diff
@@ -0,0 +1,193 @@
+import { Express } from 'express'
+import { EpochKeyProof } from '@unirep/circuits'
+import { errorHandler } from '../middleware'
+import { UnirepSocialSynchronizer } from '../synchornizer'
+import { VoteAction } from '../types'
+
+export default (
+    app: Express,
+    db: DB,
+    synchronizer: UnirepSocialSynchronizer
+) => {
+    app.post(
+        '/api/vote',
+        errorHandler(async (req, res, next) => {
+            await Vote(req, res, db, synchronizer)
+        })
+    )
+
+    // TODO: add get api to fetch all votes of post Id for frontend to check if user voted or not
+}
+
+/**
+ * Use this function to help us find out which vote action the user
+ * is going to do
+ *
+ * @param vote true for upvote action, false for downvote action
+ * @param findVote the vote data, if it's undefined then it's the first time to vote
+ * @returns voteAction or error if it's wrong action
+ */
+function selectVoteAction(vote: boolean, findVote: any): VoteAction | Error {
+    let voteAction: VoteAction
+    if (findVote) {
+        if (vote && findVote.upVote) {
+            voteAction = VoteAction.CANCEL_UPVOTE
+        } else if (!vote && findVote.downVote) {
+            voteAction = VoteAction.CANCEL_DOWNVOTE
+        } else {
+            return new Error('wrong vote action')
+        }
+    } else {
+        if (vote) {
+            voteAction = VoteAction.UPVOTE
+        } else {
+            voteAction = VoteAction.DOWNVOTE
+        }
+    }
+    return voteAction
+}
+
+/**
+ * Since the action of the vote will effect the post as well,
+ * we should use transactionDB to execute these two transactions at once
+ * to make sure the integrity
+ *
+ * @param db SQLiteConnector
+ * @param epochKeyProof construct epochKeyProof from zkp
+ * @param post the post that the user is about to vote
+ * @param voteAction the vote action of the user
+ */
+async function exeuteTxs(
+    db: DB,
+    epochKey: string,
+    post: any,
+    voteAction: VoteAction
+): Promise<void> {
+    const _id = post._id
+
+    // using upsert, so left update blank for update statement
+    const voteStatement = {
+        where: {
+            postId: _id,
+            epochKey: epochKey,
+        },
+        create: {
+            postId: _id,
+            epochKey: epochKey,
+            upVote: false,
+            downVote: false,
+        },
+        update: {},
+    }
+
+    // only modify the upCount and downCount
+    const postStatement = {
+        where: {
+            _id: _id,
+        },
+        update: {
+            upCount: post.upCount,
+            downCount: post.downCount,
+        },
+    }
+    switch (voteAction) {
+        case VoteAction.UPVOTE:
+            voteStatement.create.upVote = true
+            postStatement.update.upCount += 1
+            break
+        case VoteAction.DOWNVOTE:
+            voteStatement.create.downVote = true
+            postStatement.update.downCount += 1
+            break
+        case VoteAction.CANCEL_UPVOTE:
```

Comment:
For the case of CANCEL_UPVOTE and CANCEL_DOWNVOTE, the whole vote entry should be deleted or the selectVoteAction part will return error if we vote again (i.e., upvote -> cancel upvote -> upvote again -> error)

Reply:
Or we should modify the `selectVoteAction` function?

Reply:
Yeah, that's problem may happen, I modify `selectVoteAction` to `verifyVoteAction` to check the action is valid or not first, then execute the transactions afterwards. 

within `executeTxs`, there's a createVote flag for the transactionDB to know it should create/delete the vote.

Reply:
That's great!!!